src/dsp_tools/commands/excel2json/new_lists.py

```diff
@@ -12,33 +12,105 @@
+    ids = []
+    for sheets in excel_dfs.values():
+        for df in sheets.values():
+            ids.extend(df["id"].tolist())
+
+    id_df = pd.DataFrame({"id": ids})
+    if (duplicates := id_df["id"].duplicated(keep=False)).any():
+        id_set = set(id_df["id"][duplicates].to_list())
+        return _resolve_duplicates_in_all_excel(id_set, excel_dfs)
+    return excel_dfs
+
+
+def _resolve_duplicates_in_all_excel(
+    id_set: set[str], excel_dfs: dict[str, dict[str, pd.DataFrame]]
+) -> dict[str, dict[str, pd.DataFrame]]:
+    for filename, sheets in excel_dfs.items():
+        for sheet, df in sheets.items():
+            preferred_lang = _get_preferred_language(df.columns)
+            for i, row in df.iterrows():
+                if row["id"] in id_set:
+                    df.at[i, "id"] = _construct_non_duplicate_id_string(df.iloc[int(str(i))], preferred_lang)
+    return excel_dfs
+
+
+def _fill_one_df_auto_id_resolve_duplicates(df: pd.DataFrame, preferred_language: str) -> pd.DataFrame:
     if "ID (optional)" not in df.columns:
         df["ID (optional)"] = pd.NA
-    df = _fill_auto_id_column(df, preferred_language)
+    df = _fill_one_df_auto_id(df, preferred_language)
     df["id"] = df["ID (optional)"].fillna(df["auto_id"])
-    df = _fill_parent_id(df, preferred_language)
+    df = _resolve_duplicate_custom_and_auto_id(df, preferred_language)
+    return df
+
+
+def _resolve_duplicate_custom_and_auto_id(df: pd.DataFrame, preferred_language: str) -> pd.DataFrame:
+    if (duplicate_filter := df["id"].duplicated(keep=False)).any():
+        for i in duplicate_filter.index[duplicate_filter]:
+            if pd.isna(df.at[i, "ID (optional)"]):
+                df.loc[i, "id"] = _construct_non_duplicate_id_string(df.iloc[i], preferred_language)
     return df
 
 
-def _fill_auto_id_column(df: pd.DataFrame, preferred_language: str) -> pd.DataFrame:
+def _fill_one_df_auto_id(df: pd.DataFrame, preferred_language: str) -> pd.DataFrame:
     """For every node without manual ID, take the label of the preferred language as ID."""
     df["auto_id"] = pd.NA
     if not df["ID (optional)"].isna().any():
         return df
     if pd.isna(df.at[0, "ID (optional)"]):
-        df.at[0, "auto_id"] = df.at[0, f"{preferred_language}_list"]
+        df.loc[0, "auto_id"] = df.at[0, f"{preferred_language}_list"]
     columns = sorted(_get_columns_of_preferred_lang(df.columns, preferred_language, r"\d+"), reverse=True)
     for i, row in df.iterrows():
         if pd.isna(row["ID (optional)"]):
             for col in columns:
                 if pd.notna(row[col]):
                     df.at[i, "auto_id"] = row[col]
                     break
+    df = _resolve_one_df_auto_duplicate_ids(df, preferred_language)
     return df
 
 
-def _fill_parent_id(df: pd.DataFrame, preferred_language: str) -> pd.DataFrame:
+def _resolve_one_df_auto_duplicate_ids(df: pd.DataFrame, preferred_language: str) -> pd.DataFrame:
```

Comment:
This seems as if you had put all words into a bag, and then shaked and mixed them ;-)

Renaming suggestion: _resolve_duplicate_auto_ids_of_one_df

Perhaps, the "of one df" could even be omitted.

Reply:
I tried to include in the function name if it did something for one df, for one excel or all excels... so I understand that they are ugly, but I focused on putting all the words in

Reply:
My concern was not so much that the name contains too many words. Rather that they are mixed up, without a syntactic coherence.